### PR TITLE
Add make_unique

### LINKED ISSUE #14169: Add make_unique
std::make_unique was added in c++14 to make it safer to construct unique_ptr<>. We should add a lightweight wrapper that uses this when available or our implementation when it is not.

### PR DESCRIPTION
**Pull request by quantumsteve**
This fixes #14169.

Adds Mantid::Kernel::make_unique, using the standard library if available, while falling back to a [reference implementation](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3656.htm) for C++11. I manually expanded the variadic template for MSVC 2012.

more information about [std::make_unique](http://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique)

No release notes.

Testing: I added a few sanity checks to the unit tests. Code review, ideally from someone familiar with `std::make_unique`

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
I compared this implementation with the one by gcc (specifically `/usr/include/c++/4.9/bits/unique_ptr.h`). I spotted only two differences (apart from variable names, etc., and not counting the MSVC part, of course):

- The struct `_Unique_if` for known for known bounds contains `struct __invalid_type {};` instead of `typedef void _Known_bound;`
- The `make_unique` functions are declared inline.

Apart from that, it looks equivalent. Not sure if those differences are relevant. In lack of better suggestions, does it make sense to just go with the way gcc does it?

In any case, given the complexity of reading code like this, maybe some else should review this as well?

**Comment by peterfpeterson** [comment]
I'm willing to give it a :+1:. Like Simon, I'd appreciate more eyeballs. @martyngigg ?

File: Framework/Kernel/inc/MantidKernel/make_unique.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+make_unique(Arg1 param1, Arg2 param2, Arg3 param3, Arg4 param4) = delete;
+template <class T, class Arg1, class Arg2, class Arg3, class Arg4, class Arg5>
+inline typename _Unique_if<T>::_Known_bound
+make_unique(Arg1 param1, Arg2 param2, Arg3 param3, Arg4 param4,
+            Arg5 param5) = delete;
+#endif // MSVC
+#endif // __cplusplus == 201402L
+}
+}
+#endif /* defined(__SpinWaveGenie_Memory__) */
```
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/make_unique.h]: Just for completeness could we remove the reference to Spin Wave Genie?

**Comment by martyngigg** [comment]
Only a few comments around cleaning up non-code related things and then I'm happy to merge.

**Comment by quantumsteve** [comment]
I removed the extraneous comment. Is this now ready to merge?

**Comment by martyngigg** [comment]
Looks good to me.

**Pull request merged**